object_gomap_test.go

```diff
@@ -326,3 +328,142 @@ func TestGoMapUnicode(t *testing.T) {
 		t.Fatalf("Unexpected value: %v", res)
 	}
 }
+
+func TestGoMapMemUsage(t *testing.T) {
+	vm := New()
+	tests := []struct {
+		name           string
+		val            *objectGoMapSimple
+		threshold      int
+		expectedMem    uint64
+		expectedNewMem uint64
+		errExpected    error
+	}{
+		{
+			name:      "should account for each key value pair given a non-empty object",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test0": valueInt(99),
+					"test1": valueInt(99),
+					"test2": valueInt(99),
+					"test3": valueInt(99),
+				},
+			},
+			// overhead + len("testN") + value
+			expectedMem: SizeEmptyStruct + (5+SizeInt)*4,
+			// overhead + len("testN") with string overhead + value
+			expectedNewMem: SizeEmptyStruct + ((5+SizeString)+SizeInt)*4,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for each key value pair given map with a nil value",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": nil,
+				},
+			},
+			// overhead + len("test") + null
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct,
+			// overhead + len("test") with string overhead + null
+			expectedNewMem: SizeEmptyStruct + (4 + SizeString) + SizeEmptyStruct,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for nested parsed map",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": &objectGoMapSimple{ // <- this is treated as a objectGoMapReflect
+						baseObject: baseObject{
+							val: &Object{runtime: vm},
+						},
+						data: map[string]interface{}{
+							"subTest1": valueInt(99),
+							"subTest2": valueInt(99),
+							"subTest3": valueInt(99),
+							"subTest4": valueInt(99),
+						},
+					},
+				},
+			},
+			// overhead + len("test") + reflectedMap
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct,
+			// overhead + len("test") with string overhead + reflectedMap
+			expectedNewMem: SizeEmptyStruct + 4 + SizeString + SizeEmptyStruct,
+			errExpected:    nil,
+		},
```

Comment:
[Q] is this test case even possible/valid?

Reply:
Not really, this type is only generated when as a goja user I call ToValue on a map[string]interface{} but the type is otherwise not exposed.

Reply:
I'm OK if you only want to focus on goMapSimple and goSlice with this PR and we can do the rest in a dedicated ticket.